### PR TITLE
ENG-479: feat(mllp-server): add timezone conversion - ny only

### PR DESCRIPTION
**Pull request by lucasdellabella**
Issues:

- https://linear.app/metriport/issue/ENG-479

### Dependencies

- Upstream: https://github.com/metriport/metriport-internal/pull/3053

### Description

This adds timezone handling code to the mllp server and adjusts the timezones coming in specifically for hixny.

### Testing

Unit tests.

### Release Plan

- [ ] Merge this

## Summary by CodeRabbit

- **New Features**
  - Incoming HL7 messages are now automatically normalized to UTC based on the sending application's timezone, improving consistency across different partners.
  - Added functionality to extract sending application identifiers from HL7 messages for timezone mapping.
  - Introduced timezone-aware datetime conversion for HL7 message fields, supporting various datetime formats and offsets.
- **Bug Fixes**
  - Datetime fields in HL7 messages are reliably converted to UTC, including handling of various segments and edge cases.
- **Tests**
  - Added comprehensive tests to verify HL7 datetime conversion, including multiple segments, timezones, and error handling.
- **Chores**
  - Configuration updates now require timezone information for each HIE, ensuring accurate timezone mapping for message normalization.
  - Environment variables now include a JSON dictionary mapping sending applications to timezones for runtime use.

### CODE REVIEW

File: packages/core/src/external/hl7-notification/__tests__/datetime.test.ts [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+    try {
+      return testFn();
+    } finally {
+      // Restore original console methods
+      console.log = originalLog;
+      console.warn = originalWarn;
+      console.error = originalError;
+    }
+  };
```
[the remark this answers is not in the document]
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/__tests__/datetime.test.ts:26]: this is a simple testing util, and in this case an arrow function is good

File: packages/core/src/external/hl7-notification/datetime.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+function utcifyHl7Components(
+  message: Hl7Message,
+  segmentName: string,
+  fieldIndex: number,
+  componentIndex: number
+): Hl7Message {
+  const { log } = out("utcifyHl7Components");
+  const hiePartner = getSendingApplication(message) ?? "Unknown HIE";
```
Comment (RamilGaripov) [inline packages/core/src/external/hl7-notification/datetime.ts]: why are we doing this on every component? i'd expect us to just get the timezone offset from the message and then go and update the message

File: packages/core/src/external/hl7-notification/hie-timezone.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+export const hieTimezones = {
```
Comment (RamilGaripov) [inline packages/core/src/external/hl7-notification/hie-timezone.ts]: i dont think we should expose a map of ADT providers here like this. Let's add an optional field on the HieConfig env config to specify their timeline. That'll be much cleaner and easier to maintain imo

File: packages/core/src/external/hl7-notification/datetime.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+   * TODO: Properly timezoneify the message using getHieTimezone, which requires making response format requests over email
+   * Temporary to get Hixny ADTs correct for now.
+   * */
+  const resolvedTimezone = hiePartner !== "HIXNY" ? getHieTimezone(hiePartner) : "America/New_York";
+
+  const updatedSegments = message.segments.map(segment => {
+    if (segment.name !== segmentName) return segment;
+
+    let component = segment.getComponent(fieldIndex, componentIndex)?.trim();
+    const noComponent = !component || component === "";
```
Comment (RamilGaripov) [inline packages/core/src/external/hl7-notification/datetime.ts]: we use `isSomething` for booleans. Maybe `isEmptyComponent`

File: packages/core/src/external/hl7-notification/datetime.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    if (noComponent) return segment;
+
+    component = stripTimezoneOffset(component, segmentName, fieldIndex);
+    component = stripSubseconds(component);
+
+    let componentUtc;
+    try {
+      componentUtc = buildDayjsTz(component, resolvedTimezone).format("YYYYMMDDHHmmss");
+    } catch (e) {
+      log(`Error UTCifying component ${component} in segment ${segmentName}`, e);
```
Comment (RamilGaripov) [inline packages/core/src/external/hl7-notification/datetime.ts]: we follow this pattern:
```
const msg = "Error UTCfying...";
log(`${msg}, error - ${errorToString(e)}`);
capture.error(msg...
```

File: packages/core/src/external/hl7-notification/datetime.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+ *
+ * Doing this for now to simplify parsing, as there aren't consistently followed formats,
+ * and the vast majority of the time no timezone offset is present.
+ */
+function stripTimezoneOffset(component: string, segmentName: string, fieldIndex: number): string {
+  const timezoneMatch = component.match(/^([\d.]+)([+-]\d{4})$/);
+  if (timezoneMatch) {
+    const [, dateTimePart, offsetPart] = timezoneMatch;
+    if (dateTimePart && offsetPart) {
+      capture.message(
```
Comment (RamilGaripov) [inline packages/core/src/external/hl7-notification/datetime.ts]: if we send something to sentry, we should also log it

File: packages/core/src/external/hl7-notification/datetime.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        }
+      );
+      return dateTimePart;
+    }
+  }
+  return component;
+}
+
+/**
+ * Strips subsecond units (e.g., .123) from HL7 datetime string
```
Comment (RamilGaripov) [inline packages/core/src/external/hl7-notification/datetime.ts]: nit: second decimals?

File: packages/core/src/external/hl7-notification/datetime.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+ *
+ * This is done to simplify parsing, as there aren't consistently followed formats,
+ * and the vast majority of the time no subsecond units are present.
```
Comment (RamilGaripov) [inline packages/core/src/external/hl7-notification/datetime.ts]: ditto

File: packages/core/src/external/hl7-notification/datetime.ts [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+ */
+function stripSubseconds(component: string): string {
```
Comment (RamilGaripov) [inline packages/core/src/external/hl7-notification/datetime.ts]: ditto

**Review by RamilGaripov — CHANGES_REQUESTED** [review]

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 2 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: yep that makes sense. I felt like it made the code messier and theres no perf impact, but yeah

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 4 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: addressed

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 5 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: addressed

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 6 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: good point

File: packages/core/src/external/hl7-notification/hie-timezone.ts (hunk 3 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/hie-timezone.ts]: cool, makes a ton of sense

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 6 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: addressed

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 5 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: addressed

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 4 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: addressed

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 2 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: addressed

File: packages/core/src/external/hl7-notification/hie-timezone.ts (hunk 3 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/hie-timezone.ts]: opted to just rename and adjust tests, its not actively doing anything right now, and this way we can fix the CCP stuff asap. This is two-way door code, we'll prob just delete it later if we want to secret-ify stuff.

File: packages/core/src/external/hl7-notification/datetime.ts [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    if (isComponentEmpty) return segment;
+
+    component = stripTimezoneOffset(component, segmentName, fieldIndex);
+    component = stripSubseconds(component);
+
+    let componentUtc;
+    try {
+      componentUtc = buildDayjsTz(component, timezone).format("YYYYMMDDHHmmss");
+    } catch (e) {
+      const msg = `Error UTCifying component ${component} in segment ${segmentName}`;
```
Comment (RamilGaripov) [inline packages/core/src/external/hl7-notification/datetime.ts]: We dont want dynamic messages for the Sentry. Keep msg a static string, and any additional details go to the logs + as extra on capture

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 9 above)
Comment (RamilGaripov) [reply packages/core/src/external/hl7-notification/datetime.ts]: why `subsecond decimals`? decimals ARE `sub` 😆 

can even call this `stripMilliseconds`, no? 😂

Comment (RamilGaripov) [reply packages/core/src/external/hl7-notification/datetime.ts]: ... why didnt I think of this the first time? ^

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 10 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: good point - I updated this and removed the component, but want to keep the segmentName - I think this is an exception where I want to break the dynamic item rule, as imo it will be helpful to have different segments to each throw their own distinct error, and the cardinality of possible errors is not like 1 million but ~5

File: packages/core/src/external/hl7-notification/hie-timezone.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+  HealthConnectTexas: "America/Chicago",
+  Hixny: "America/New_York",
+  HieTexasPcc: "America/Chicago",
+  SacValleyMS: "America/Los_Angeles",
+};
+
+export function getHieTimezone(partnerName: string): string {
+  return hieTimezones[partnerName as keyof typeof hieTimezones];
+}
```
[the remark this answers is not in the document]
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/hie-timezone.ts]: pass for now. will update later if deemed necessary

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 9 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: hahahaha yeah i was a bit trigger happy with rename. went with decimals because its actually not always millis we're stripping. sometimes its hundreths or tenths, whatever it is we just wanna wipe the period and following values.

File: packages/core/src/external/hl7-notification/datetime.ts [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+export function utcifyHl7Message(message: Hl7Message): Hl7Message {
+  const sendingApplication = getSendingApplication(message);
+  const hiePartner = sendingApplication ?? "Unknown HIE";
+  capture.setExtra({ sendingApplication });
+
+  /**
+   * TODO: Properly timezoneify the message using getHieTimezone, which requires making response format requests over email
+   * Temporary to get NewYorkHie ADTs correct for now.
+   * */
+  const resolvedTimezone = hiePartner !== "HIXNY" ? getHieTimezone(hiePartner) : "America/New_York";
```
Comment (leite08) [inline packages/core/src/external/hl7-notification/datetime.ts]: let's use boolean in the truthly format when possible - in this case, if partner is HXNY do this, otherwise that

or, even better, build a map w/ the parner and the tz :)

Comment (leite08) [reply packages/core/src/external/hl7-notification/datetime.ts]: BTW, can use offset instead of named TZs? Simpler, doesn't require people to think, remember, google it

File: packages/core/src/external/hl7-notification/datetime.ts [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+      debug(msg);
+      capture.message(msg, {
+        extra: { segmentName, fieldIndex, originalComponent: component, offset: offsetPart },
+        level: "warning",
+      });
+      return dateTimePart;
+    }
+  }
+  return component;
+}
```
Comment (leite08) [inline packages/core/src/external/hl7-notification/datetime.ts]: We should adjust the datetime to UTC if it has an offset before returning it. Even though most times they don't have it, if we build code to remove it, I think we should adjust it.

Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: cool ill do this

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 9 above)
Comment (RamilGaripov) [reply packages/core/src/external/hl7-notification/datetime.ts]: but even if it's hundreds or tenths, they're still hundreds of milliseconds lol

Comment (RamilGaripov) [reply packages/core/src/external/hl7-notification/datetime.ts]: nbad tho, im ok with decimals too

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 10 above)
Comment (RamilGaripov) [reply packages/core/src/external/hl7-notification/datetime.ts]: hmmmm.. why tho? even if they all go under the same sentry error, we should be able to see all that we need to see when we click into it

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 12 above)
Comment (RamilGaripov) [reply packages/core/src/external/hl7-notification/datetime.ts]: offset in numeric values doesnt cut it cuz it changes throughout the year based on daylight savings

Comment (RamilGaripov) [reply packages/core/src/external/hl7-notification/datetime.ts]: tho idk if the current approach actually accounts for it. I assume it does

File: packages/core/src/external/hl7-notification/datetime.ts [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+ */
+function stripDecimals(component: string): string {
+  const { debug } = out("stripDecimals");
+  const subsecondMatch = component.match(/^(\d{14})(\.\d+)$/);
+  if (subsecondMatch) {
+    const [, dateTimePart] = subsecondMatch;
+    if (dateTimePart) {
+      const msg = `Dropped decimals from HL7 datetime field`;
+      debug(msg);
+      capture.message(msg, {
```
Comment (RamilGaripov) [inline packages/core/src/external/hl7-notification/datetime.ts]: do we want this? seems like this could be very noisy?

File: packages/core/src/util/config.ts [hunk 15]
```diff
@@ -105,6 +105,14 @@ export class Config {
...
     return getEnvVarOrFail("HL7_NOTIFICATION_QUEUE_URL");
   }
+  static getHieTimezoneDictionary(): Record<string, string> {
+    const hieTimezoneDictionary = getEnvVarOrFail("HIE_TIMEZONE_DICTIONARY");
+    try {
+      return JSON.parse(hieTimezoneDictionary) as Record<string, string>;
+    } catch (error) {
+      throw new Error(`Failed to parse HIE_TIMEZONE_DICTIONARY env var: ${error}`);
+    }
+  }
```
Comment (RamilGaripov) [inline packages/core/src/util/config.ts]: i dont think this file should contain any logic whatsoever. `getHieTimezoneDictionary` should just return the dictionary, and you can parse it with a separate helper function

File: packages/mllp-server/.env.example [hunk 16]
```diff
@@ -6,6 +6,8 @@ HL7_INCOMING_MESSAGE_BUCKET_NAME=your-incoming-bucket-name
 HL7_OUTGOING_MESSAGE_BUCKET_NAME=your-outgoing-bucket-name
 HL7_CONVERSION_BUCKET_NAME=metriport-hl7-conversion-staging
 HL7_NOTIFICATION_QUEUE_URL=your-queue-url
+HIE_TIMEZONE_DICTIONARY="'{"myHie":"Some/Timezone"}'"
```
[the remark this answers is not in the document]
Comment (RamilGaripov) [reply packages/mllp-server/.env.example]: i'd prefer an example that accurately represents a timezone, like 
`"'{"myHie":"America/New_York"}'"`

File: packages/core/src/util/config.ts (hunk 15 above)
Comment (RamilGaripov) [reply packages/core/src/util/config.ts]: whatever place uses this, it should be like:
```
const hieTimezoneDictionary = getHieTimezoneDictionary();

and 

function getHieTimezoneDictionary() {
 const rawHieDictionary = Config.getHieTimezoneDictionary();
//  processing
}

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 10 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: But why not? Some segments are more important than others. I'd like to know if we're getting a sentry failure for PV2 or MSH. This rule exists because if we put dynamic data in the message the events don't get properly tied together in sentry, and to avoid messiness and maintain searchability of an error message. I made your suggested change, and retaining segment name still allows all that.

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 12 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: yep IANA timezones (America/New_York) adjust based on daylight savings

File: packages/core/src/external/hl7-notification/datetime.ts [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+
+    return date.format("YYYYMMDDHHmmss");
+  } catch (e) {
+    const msg = `Error adjusting timezone offset in HL7 datetime field`;
+    debug(`${msg}, error - ${errorToString(e)}`);
+    capture.error(msg, {
+      extra: { segmentName, fieldIndex, originalComponent: `${dateTimePart}${offsetPart}` },
+    });
+    return dateTimePart;
+  }
```
[the remark this answers is not in the document]
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts:161]: that doesnt exist

File: packages/core/src/external/hl7-notification/datetime.ts (hunk 14 above)
Comment (lucasdellabella) [reply packages/core/src/external/hl7-notification/datetime.ts]: yep good point, dont think we need to know about this if it happens

File: packages/core/src/util/config.ts (hunk 15 above)
Comment (lucasdellabella) [reply packages/core/src/util/config.ts]: I actually disagree - the job of these functions are to get the value from the env file for the app to use. They are essentially a wrapper lib around our env vars. Every value in this file just _happens_ to be a string because we almost exclusively store strings in env vars. If these values are any other value, they should be parsed and made available to the application as the appropriate type. If the parsing fails, they should fail just like they would if the env var wasn't there at all.

**Review by RamilGaripov — APPROVED** [review]

File: packages/infra/lib/hl7-notification-stack/mllp.ts [hunk 18]
```diff
@@ -58,12 +59,23 @@ const setupNlb = (identifier: string, vpc: ec2.Vpc, nlb: elbv2.NetworkLoadBalanc
...
+const createHieTimezoneMap = (hieConfigs: Record<string, HieConfig>) => {
+  return Object.values(hieConfigs).reduce((acc, item) => {
+    const key = item.identifierInMshSegment;
+    const val = item.timezone;
+    if (key && val) {
+      acc[key] = val;
+    }
+    return acc;
+  }, {} as Record<string, string>);
+};
```
Comment (leite08) [inline packages/infra/lib/hl7-notification-stack/mllp.ts:71]: @coderabbitai you should've flagged this as we avoid using const/arrow functions, as indicated to you before a few times.

File: packages/mllp-server/.env.example [hunk 19]
```diff
@@ -6,6 +6,8 @@ HL7_INCOMING_MESSAGE_BUCKET_NAME=your-incoming-bucket-name
 HL7_OUTGOING_MESSAGE_BUCKET_NAME=your-outgoing-bucket-name
 HL7_CONVERSION_BUCKET_NAME=metriport-hl7-conversion-staging
 HL7_NOTIFICATION_QUEUE_URL=your-queue-url
+HIE_TIMEZONE_DICTIONARY='{myHie":"America/New_York"}'
```
[the remark this answers is not in the document]
Comment (leite08) [reply packages/mllp-server/.env.example:9]: @lucasdellabella NABD as its an example, but good to fix on a future PR

File: packages/shared/src/common/env-var.ts [hunk 20]
```diff
@@ -17,6 +17,15 @@ export const getEnvVarOrFail = (varName: string): string => {
...
 };
 
+export const getEnvVarAsRecordOrFail = (varName: string): Record<string, string> => {
+  const value = getEnvVarOrFail(varName);
+  try {
+    return JSON.parse(value) as Record<string, string>;
+  } catch (error) {
+    throw new Error(`Failed to parse ${varName} env var`);
+  }
+};
```
Comment (leite08) [inline packages/shared/src/common/env-var.ts:27]: @coderabbitai another example of const/arrow function that should have been flagged
